File: src/types/NormalizedNotification.ts
export type NotificationPriority = 'MIN' | 'LOW' | 'DEFAULT' | 'HIGH' | 'MAX';

export type NotificationType = 'GENERIC' | 'CHAT' | 'COMMERCE' | 'MEDIA' | 'PROGRESS';

export interface NotificationAction {
  id: string;
  title: string;
  type: 'BUTTON' | 'TEXT_INPUT' | 'DESTRUCTIVE';
  placeholder?: string;
  authenticationRequired?: boolean;
}

export interface ChatMessage {
  id: string;
  text: string;
  timestamp: number;
  senderId: string;
  senderName: string;
  senderAvatar?: string;
}

export interface ChatConversation {
  threadId: string;
  senderId: string;
  senderName: string;
  senderAvatar?: string;
  isGroup: boolean;
  groupName?: string;
  messages: ChatMessage[];
  unreadCount?: number;
}

export interface CommerceProgress {
  orderId: string;
  currentStep: number;
  totalSteps: number;
  stepName: string;
  progressPercent: number;
  estimatedDelivery?: string;
  statusText?: string;
  trackerUrl?: string;
  silentUpdate?: boolean;
}

export interface MediaMetadata {
  title?: string;
  artist?: string;
  album?: string;
  artworkUrl?: string;
  isPlaying?: boolean;
  durationSeconds?: number;
  playbackPositionSeconds?: number;
}

export interface NormalizedNotification {
  id: string;
  messageId: string;
  collapseKey?: string;
  threadId?: string;
  category?: string;
  title: string;
  body: string;
  subtitle?: string;
  badge?: number;
  sound?: string;
  priority: NotificationPriority;
  avatar?: string;
  imageUrl?: string;
  color?: string;
  useCustomLayout?: boolean;
  gradientColors?: string[];
  deepLink?: string;
  data: Record<string, any>;
  actions?: NotificationAction[];
  timestamp: number;
  type: NotificationType;
  commerceMetadata?: CommerceProgress;
  chatMetadata?: ChatConversation;
  mediaMetadata?: MediaMetadata;
  isRead?: boolean;
  isSuppressed?: boolean;
  silentUpdate?: boolean;
}

export interface NotificationChannel {
  id: string;
  name: string;
  description?: string;
  importance?: 'NONE' | 'MIN' | 'LOW' | 'DEFAULT' | 'HIGH' | 'MAX';
  sound?: string;
  enableVibration?: boolean;
  showBadge?: boolean;
}

export interface NotificationCategory {
  id: string;
  actions: NotificationAction[];
  hiddenPreviewsBodyPlaceholder?: string;
}

export interface RuleResult {
  action: 'PRESENT' | 'SUPPRESS' | 'SUPPRESS_AND_INJECT' | 'SILENT_UPDATE';
  modifiedNotification?: NormalizedNotification;
  reason?: string;
}

export interface NotificationRule {
  id: string;
  description?: string;
  condition: (notification: NormalizedNotification, appState: { activeChatId?: string; isForeground: boolean }) => boolean;
  action: RuleResult['action'];
}

export interface RNNOSPlugin {
  name: string;
  onInitialize?: () => void;
  onPushReceived?: (notification: NormalizedNotification) => Promise<NormalizedNotification | null>;
  onNotificationTapped?: (notification: NormalizedNotification, actionId?: string) => void;
  onNotificationDismissed?: (notificationId: string) => void;
}

export interface RNNOSConfig {
  appId: string;
  channels?: NotificationChannel[];
  categories?: NotificationCategory[];
  rules?: NotificationRule[];
  plugins?: RNNOSPlugin[];
  autoMarkReadOnScreenFocus?: boolean;
  badgeSyncEnabled?: boolean;
}

File: ios/Sources/RNNOS/Domain/Models/NormalizedNotification.swift
import Foundation

public enum NotificationPriority: String, Codable {
    case MIN, LOW, DEFAULT, HIGH, MAX
}

public enum NotificationType: String, Codable {
    case GENERIC, CHAT, COMMERCE, MEDIA, PROGRESS
}

public struct NotificationAction: Codable {
    public let id: String
    public let title: String
    public let type: String // BUTTON, TEXT_INPUT, DESTRUCTIVE
    public let placeholder: String?
    public let authenticationRequired: Bool?
}

public struct ChatMessage: Codable {
    public let id: String
    public let text: String
    public let timestamp: Double
    public let senderId: String
    public let senderName: String
    public let senderAvatar: String?
}

public struct ChatConversation: Codable {
    public let threadId: String
    public let senderId: String
    public let senderName: String
    public let senderAvatar: String?
    public let isGroup: Bool
    public let groupName: String?
    public let messages: [ChatMessage]
    public let unreadCount: Int?
}

public struct CommerceProgress: Codable {
    public let orderId: String
    public let currentStep: Int
    public let totalSteps: Int
    public let stepName: String
    public let progressPercent: Int
    public let estimatedDelivery: String?
    public let statusText: String?
    public let trackerUrl: String?
}

public struct NormalizedNotification: Codable {
    public let id: String
    public let messageId: String
    public let collapseKey: String?
    public let threadId: String?
    public let category: String?
    public let title: String
    public let body: String
    public let subtitle: String?
    public let badge: Int?
    public let sound: String?
    public let priority: NotificationPriority
    public let avatar: String?
    public let imageUrl: String?
    public let deepLink: String?
    public let data: [String: String]
    public let actions: [NotificationAction]
    public let timestamp: Double
    public let type: NotificationType
    public let commerceMetadata: CommerceProgress?
    public let chatMetadata: ChatConversation?
    public let isRead: Bool?
    public let isSuppressed: Bool?

    public static fun parse(jsonString: String) -> NormalizedNotification? {
        guard let data = jsonString.data(using: .utf8) else { return nil }
        let decoder = JSONDecoder()
        return try? decoder.decode(NormalizedNotification.self, from: data)
    }

    public func toJsonString() -> String? {
        let encoder = JSONEncoder()
        guard let data = try? encoder.encode(self) else { return nil }
        return String(data: data, encoding: .utf8)
    }
}
